daemon/daemon_cleanup.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2017 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/cilium/cilium/daemon/defaults"
+	"github.com/cilium/cilium/pkg/bpf"
+	"github.com/vishvananda/netlink"
+)
+
+const (
+	ciliumPrefix = "cilium"
+	tunnelMap    = "tunnel_endpoint_map"
+	cniConfig    = "/etc/cni/net.d/10-cilium-cni.conf"
+)
+
+func runCleanup() {
+	// Abort if the agent is running
+	if _, err := os.Stat(defaults.SockPath); !os.IsNotExist(err) {
+		fmt.Fprintf(os.Stderr, "Agent should not be running when cleaning up\n"+
+			"Found socket %s\n", defaults.SockPath)
+		os.Exit(1)
+	}
+
+	// Ask user for confirmation
+	if !cleanupConfirmed() {
+		return
+	}
+
+	// ENOENT and similiar errors are ignored. Should print all other
+	// errors seen, but continue.  So that one remove function does not
+	// prevent the reaming from running.
+	type cleanupFunc func() error
+	checks := []cleanupFunc{removeAllMaps, unmountFS, removeDirs, removeCNI, removeRoutes}
+	for _, clean := range checks {
+		if err := clean(); err != nil {
+			fmt.Fprintf(os.Stderr, "Error: %s\n", err)
+		}
+	}
+}
+
+func cleanupConfirmed() bool {
+	fmt.Printf("Destructive operation! This will remove:\n" +
+		"- all BPF maps\n" +
+		"- mounted bpffs\n" +
+		"- routes and links\n" +
+		"- library code\n" +
+		"- endpoint state\n" +
+		"- CNI configuration\n" +
+		"The command is non-revertable, do you want to contine [y/N]?")
+
+	var res string
+	fmt.Scanln(&res)
+	return res == "y"
+}
+
+func removeCNI() error {
+	err := os.Remove(cniConfig)
+	if os.IsNotExist(err) {
+		return nil
+	}
+	return err
+}
+
+func removeDirs() error {
+	dirs := []string{defaults.RuntimePath, defaults.LibraryPath}
+	for _, dir := range dirs {
+		if err := os.RemoveAll(dir); err != nil {
```

Comment:
to be precocious throw an error if `dir` is `/` before doing a `RemoveAll`

Reply:
Fixed